src/main/java/org/traccar/sms/SnsSmsClient.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2021 Anton Tananaev (anton@traccar.org)
+ * Copyright 2021 Subodh Ranadive (subodhranadive3103@gmail.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.sms;
+
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.services.sns.AmazonSNSAsync;
+import com.amazonaws.services.sns.AmazonSNSAsyncClientBuilder;
+import com.amazonaws.services.sns.model.MessageAttributeValue;
+import com.amazonaws.services.sns.model.PublishRequest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.traccar.Context;
+import org.traccar.config.Keys;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SnsSmsClient implements SmsManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SnsSmsClient.class);
+
+    private final AmazonSNSAsync snsClient;
+
+    public SnsSmsClient() {
+        if (Context.getConfig().hasKey(Keys.SMS_AWS_REGION)
+                && Context.getConfig().hasKey(Keys.SMS_AWS_ACCESS)
+                && Context.getConfig().hasKey(Keys.SMS_AWS_SECRET)) {
+            BasicAWSCredentials awsCredentials =
+                    new BasicAWSCredentials(Context.getConfig().getString(Keys.SMS_AWS_ACCESS),
+                    Context.getConfig().getString(Keys.SMS_AWS_SECRET));
+            snsClient = AmazonSNSAsyncClientBuilder.standard()
+                    .withRegion(Context.getConfig().getString(Keys.SMS_AWS_REGION))
+                    .withCredentials(new AWSStaticCredentialsProvider(awsCredentials)).build();
+        } else {
+            throw new RuntimeException("SNS Not Configured Properly. Please provide valid config.");
+        }
+    }
+
+    @Override
+    public void sendMessageSync(String destAddress, String message, boolean command) {
+        Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
+        smsAttributes.put("AWS.SNS.SMS.SenderID",
+                new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
+        smsAttributes.put("AWS.SNS.SMS.SMSType",
+                new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
+        try {
+            snsClient.publishAsync(new PublishRequest().withMessage(message)
+                    .withPhoneNumber(destAddress).withMessageAttributes(smsAttributes));
+        } catch (Exception exception) {
```

Comment:
I don't think that's how you handle asynchronous errors. This seems incorrect. Also, why is it a generic `Exception`?

Reply:
Hi @tananaev 
I have changed the generic `Exception` to `RuntimeException`.

Regarding handling of asynchronous errors, I took the reference of `publishAsync()` method of `AmazonSNSAsyncClient.java` as below 

[https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-sns/src/main/java/com/amazonaws/services/sns/AmazonSNSAsyncClient.java#L1337-L1362](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-sns/src/main/java/com/amazonaws/services/sns/AmazonSNSAsyncClient.java#L1337-L1362)

Here they are throwing the exception and as we have added catch block in our code and as a result that will be handled.

Reply:
But it is thrown in a function that is presumably executed asynchronously. You cannot catch that here.

Reply:
Hi @tananaev 

Can we use something like below:

```
Thread.setDefaultUncaughtExceptionHandler((thread, exception) ->
                LOGGER.error("SMS send failed", exception));

        new Thread(() -> {
            Map<String, MessageAttributeValue> smsAttributes = new HashMap<>();
            smsAttributes.put("AWS.SNS.SMS.SenderID",
                    new MessageAttributeValue().withStringValue("SNS").withDataType("String"));
            smsAttributes.put("AWS.SNS.SMS.SMSType",
                    new MessageAttributeValue().withStringValue("Transactional").withDataType("String"));
            snsClient.publishAsync(new PublishRequest().withMessage(message)
                    .withPhoneNumber(destAddress).withMessageAttributes(smsAttributes));
        });
```
If not, Please can you provide a small example/snippet.

Reply:
I would think that because it's already `publishAsync`, you don't need a new thread. I don't have a snippet obviously.

Reply:
Hi @tananaev 

I have added AWS `AsyncHandler` for exception handling of `publishAsync()` method.